core/src/main/java/io/seata/core/rpc/netty/ChannelManager.java

```diff
@@ -121,81 +136,99 @@ private static RpcContext buildChannelHolder(NettyPoolKey.TransactionRole client
     }
 
     /**
      * Register rm channel.
      *
-     * @param resourceManagerRequest the resource manager request
+     * @param request the resource manager request
      * @param channel                the channel
      * @throws IncompatibleVersionException the incompatible  version exception
      */
-    public static void registerRMChannel(RegisterRMRequest resourceManagerRequest, Channel channel)
-        throws IncompatibleVersionException {
-        Version.checkVersion(resourceManagerRequest.getVersion());
-        Set<String> dbkeySet = dbKeytoSet(resourceManagerRequest.getResourceIds());
-        RpcContext rpcContext;
-        if (!IDENTIFIED_CHANNELS.containsKey(channel)) {
-            rpcContext = buildChannelHolder(NettyPoolKey.TransactionRole.RMROLE, resourceManagerRequest.getVersion(),
-                resourceManagerRequest.getApplicationId(), resourceManagerRequest.getTransactionServiceGroup(),
-                resourceManagerRequest.getResourceIds(), channel);
-            rpcContext.holdInIdentifiedChannels(IDENTIFIED_CHANNELS);
-        } else {
-            rpcContext = IDENTIFIED_CHANNELS.get(channel);
-            rpcContext.addResources(dbkeySet);
-        }
-        if (dbkeySet == null || dbkeySet.isEmpty()) { return; }
-        for (String resourceId : dbkeySet) {
-            String clientIp;
-            ConcurrentMap<Integer, RpcContext> portMap = CollectionUtils.computeIfAbsent(RM_CHANNELS, resourceId, key -> new ConcurrentHashMap<>())
-                    .computeIfAbsent(resourceManagerRequest.getApplicationId(), key -> new ConcurrentHashMap<>())
-                    .computeIfAbsent(clientIp = ChannelUtil.getClientIpFromChannel(channel), key -> new ConcurrentHashMap<>());
+    public static void registerRMChannel(RegisterRMRequest request, Channel channel)
+            throws IncompatibleVersionException {
+        Version.checkVersion(request.getVersion());
+
+        String resourceIds = request.getResourceIds();
+        Set<String> idSet = dbKeytoSet(resourceIds);
+
+        String typeName = request.getExtraData();
+        BranchType type = getBranchType(typeName);
+
+        RpcContext rpcContext = IDENTIFIED_CHANNELS.computeIfAbsent(channel, chan -> buildRpcContext(chan, request));
+        rpcContext.addResources(idSet);
+
+        String appId = request.getApplicationId();
+        String clientIp = ChannelUtil.getClientIpFromChannel(channel);
+        Integer clientPort = ChannelUtil.getClientPortFromChannel(channel);
+
+        idSet.forEach(resourceId -> {
+            addBranchTypeToResource(resourceId, type);
 
+            ConcurrentMap<Integer, RpcContext> portMap = getRmPortMap(resourceId, appId, clientIp);
+            portMap.put(clientPort, rpcContext);
             rpcContext.holdInResourceManagerChannels(resourceId, portMap);
-            updateChannelsResource(resourceId, clientIp, resourceManagerRequest.getApplicationId());
+        });
```

Comment:
Multiple RM channels and resourceId on the same client.
ip: ip1   channels: channel1, channel2    resourceId: resourceId1, resourceId2.

1.  resourceId1 init -> create channel1  send  RegisterRMRequest to TC.
2. resourceId2 init
3. create channel2  send resourceId1 & resourceId2 to TC
4. begin global transaction with resourceId2 through channel2
5.  channel2 inactive 
6. how use channel1 to continue the resourceId2's transaction?